### PR TITLE
processChunk in toInputStream

### PR DESCRIPTION
**Pull request by sideeffffect**

### CODE REVIEW

**Comment by iravid** [comment]
`StreamChunkSpec` is timing out locally for me too, so this looks like an actual problem.

**Comment by iravid** [comment]
It appears that the `toInputStream` is hanging now

**Comment by sideeffffect** [comment]
I suspect that `processChunk` is at fault here, but I don't really understand what could be the issue with it... I tried to rework it by wrapping it all in `Managed.effectTotal`
let's hope it helps

**Comment by iravid** [comment]
Hi @sideeffffect, any luck with this?

**Comment by sideeffffect** [comment]
@iravid sadly not, I have no idea why a simple, type-preserving, reorganization of the code suddenly causes tests to hang and whatnot :crying_cat_face:

**Comment by iravid** [comment]
@sideeffffect 

The counter is never increased in `processChunk`. This does the trick after the while loop:
```scala
          val elem = chunk(counter)
          counter += 1
          elem
```

File: streams/shared/src/main/scala/zio/stream/StreamEffectChunk.scala [hunk 1]
```diff
@@ -90,18 +90,25 @@ private[stream] class StreamEffectChunk[-R, +E, +A](override val chunks: StreamE
...
-          counter = 0
+      Managed.effectTotal {
+        var counter         = 0
+        var chunk: Chunk[A] = Chunk.empty
+        def pull(): A = {
+          while (counter >= chunk.length) {
+            chunk = thunk()
+            counter = 0
+          }
+          chunk(counter)
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/StreamEffectChunk.scala]:
```suggestion
          val elem = chunk(counter)
          counter += 1
          elem
```

File: streams/shared/src/main/scala/zio/stream/StreamEffectChunk.scala [hunk 2]
```diff
@@ -90,18 +90,25 @@ private[stream] class StreamEffectChunk[-R, +E, +A](override val chunks: StreamE
...
         }
-        chunk(counter)
+        () => pull()
       }
-      ZManaged.fromEffect(ZIO.succeed(() => pull()))
+    }
+
+  final def flatten: StreamEffect[R, E, A] =
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/StreamEffectChunk.scala]: This is `flattenChunks` and can override the method defined in `StreamChunk`.

File: streams/shared/src/main/scala/zio/stream/StreamEffectChunk.scala (hunk 1 above)
Comment (sideeffffect) [reply streams/shared/src/main/scala/zio/stream/StreamEffectChunk.scala]: :man_facepalming: of course! thank you

**Pull request merged**